chain/syncer.go

```diff
@@ -423,17 +340,53 @@ func (syncer *Syncer) HandleNewTipset(ctx context.Context, tsKey types.TipSetKey
 	defer syncer.mu.Unlock()
 
 	// If the store already has this tipset then the syncer is finished.
-	if _, err := syncer.chainStore.GetTipSet(tsKey); err == nil {
+	if syncer.chainStore.HasTipSetAndState(ctx, tsKey.String()) {
 		return nil
 	}
 
-	// Walk the chain given by the input blocks back to a known tipset in
-	// the store. This is the only code that may go to the network to
-	// resolve cids to blocks.
-	chain, err := syncer.collectChain(ctx, tsKey)
-	if err != nil {
-		return err
+	syncDoneCb := func(t types.TipSet) (bool, error) {
+		return syncer.chainStore.HasTipSetAndState(ctx, t.String()), nil
 	}
+
+	catchDoneCb := func(t types.TipSet) (bool, error) {
+		// Only check if the head we are fetching exceeds finality
+		if t.Key().Equals(tsKey) {
+			ok, err := syncer.exceedsFinalityLimit(t)
+			if err != nil {
+				return true, err
+			}
+			if ok {
+				return true, ErrNewChainTooLong
+			}
+		}
+		if syncer.chainStore.HasTipSetAndState(ctx, t.String()) {
+			return true, nil
+		}
+		return false, nil
+	}
+
+	var chain []types.TipSet
+	switch syncer.syncMode {
+	case Syncing:
+		chain, err = syncer.fetcher.FetchTipSets(ctx, tsKey, syncDoneCb)
+		if err != nil {
+			return err
+		}
+	case CaughtUp:
+		// we cancel the fetch during caught up mode after a single blocktime.
```

Comment:
This comment will rot if `blkWaitTime` changes. 

`// Time out if this bounded fetch doesn't complete quickly`